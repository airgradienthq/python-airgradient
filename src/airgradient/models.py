"""Models for AirGradient."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum

from mashumaro import field_options
from mashumaro.mixins.orjson import DataClassORJSONMixin


@dataclass
class Measures(DataClassORJSONMixin):
    """Measures model."""

    signal_strength: int = field(metadata=field_options(alias="wifi"))
    serial_number: str = field(metadata=field_options(alias="serialno"))
    boot_time: int = field(metadata=field_options(alias="boot"))
    firmware_version: str = field(metadata=field_options(alias="firmwareVersion"))
    model: str = field(metadata=field_options(alias="fwMode"))
    rco2: int | None = None
    pm01: int | None = None
    pm02: int | None = None
    pm10: int | None = None
    total_volatile_organic_component_index: int | None = field(
        default=None, metadata=field_options(alias="tvocIndex")
    )
    raw_total_volatile_organic_component: int | None = field(
        default=None, metadata=field_options(alias="tvoc_raw")
    )
    pm003_count: int | None = field(
        default=None, metadata=field_options(alias="pm003Count")
    )
    nitrogen_index: int | None = field(
        default=None, metadata=field_options(alias="noxIndex")
    )
    raw_nitrogen: int | None = field(
        default=None, metadata=field_options(alias="nox_raw")
    )
    ambient_temperature: float | None = field(
        default=None, metadata=field_options(alias="atmp")
    )
    relative_humidity: float | None = field(
        default=None, metadata=field_options(alias="rhum")
    )


class PmStandard(StrEnum):
    """PM standard model."""

    UGM3 = "ugm3"
    USAQI = "usaqi"


class TemperatureUnit(StrEnum):
    """Temperature unit model."""

    CELSIUS = "c"
    FAHRENHEIT = "f"


class ConfigurationControl(StrEnum):
    """Configuration control model."""

    CLOUD = "cloud"
    LOCAL = "local"
    BOTH = "both"


class LedBarMode(StrEnum):
    """LED bar mode."""

    OFF = "off"
    CO2 = "co2"
    PM = "pm"


@dataclass
class Config(DataClassORJSONMixin):
    """Config model."""

    country: str
    pm_standard: PmStandard = field(metadata=field_options(alias="pmStandard"))
    led_bar_mode: LedBarMode = field(metadata=field_options(alias="ledBarMode"))
    co2_automatic_baseline_calibration_days: int = field(
        metadata=field_options(alias="abcDays")
    )
    temperature_unit: TemperatureUnit = field(
        metadata=field_options(alias="temperatureUnit")
    )
    configuration_control: ConfigurationControl = field(
        metadata=field_options(alias="configurationControl")
    )
    post_data_to_airgradient: bool = field(
        metadata=field_options(alias="postDataToAirGradient")
    )
