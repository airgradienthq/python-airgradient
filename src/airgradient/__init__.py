"""Asynchronous Python client for AirGradient."""


__all__ = [
]
