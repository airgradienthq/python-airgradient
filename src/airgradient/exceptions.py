"""Asynchronous Python client for AirGradient."""


class AirGradientError(Exception):
    """Generic exception."""


class AirGradientConnectionError(AirGradientError):
    """AirGradient connection exception."""
