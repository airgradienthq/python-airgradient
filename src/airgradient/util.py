"""Utility functions for AirGradient."""

from __future__ import annotations


def get_model_name(model_id: str) -> str | None:
    """Get model name from identifier."""
    if model_id.startswith("I-9PSL"):
        return "AirGradient ONE"
    if model_id.startswith("O-1"):
        return "AirGradient Open Air"
    if "DIY" in model_id:
        return "AirGradient DIY"
    return None
