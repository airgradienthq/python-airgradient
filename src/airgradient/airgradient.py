"""Asynchronous Python client for AirGradient."""

from __future__ import annotations

from asyncio import timeout
from dataclasses import dataclass
from importlib import metadata
from typing import TYPE_CHECKING, Any

from aiohttp import ClientSession
from aiohttp.hdrs import METH_GET, METH_PUT
from yarl import URL

from .exceptions import AirGradientConnectionError
from .models import (
    Config,
    ConfigurationControl,
    LedBarMode,
    Measures,
    PmStandard,
    TemperatureUnit,
)

if TYPE_CHECKING:
    from typing_extensions import Self


VERSION = metadata.version(__package__)


@dataclass
class AirGradientClient:
    """Main class for handling connections with AirGradient."""

    host: str
    session: ClientSession | None = None
    request_timeout: int = 10
    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        data: dict[str, Any] | None = None,
    ) -> str:
        """Handle a request to AirGradient."""
        url = URL.build(scheme="http", host=self.host).joinpath(uri)

        headers = {
            "User-Agent": f"PythonAirGradient/{VERSION}",
            "Accept": "application/json",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        async with timeout(self.request_timeout):
            response = await self.session.request(
                method,
                url,
                headers=headers,
                json=data,
            )

        if response.status != 200:
            content_type = response.headers.get("Content-Type", "")
            text = await response.text()
            msg = "Unexpected response from AirGradient"
            raise AirGradientConnectionError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        return await response.text()

    async def get_current_measures(self) -> Measures:
        """Get current measures from AirGradient."""
        response = await self._request("measures/current")
        return Measures.from_json(response)

    async def get_config(self) -> Config:
        """Get config from AirGradient device."""
        response = await self._request("config")
        return Config.from_json(response)

    async def _set_config(self, field: str, value: Any) -> None:
        """Set config on AirGradient device."""
        await self._request("config", method=METH_PUT, data={field: value})

    async def set_pm_standard(self, pm_standard: PmStandard) -> None:
        """Set PM standard on AirGradient device."""
        await self._set_config("pmStandard", pm_standard)

    async def set_temperature_unit(self, temperature_unit: TemperatureUnit) -> None:
        """Set temperature unit on AirGradient device."""
        await self._set_config("temperatureUnit", temperature_unit)

    async def set_configuration_control(
        self, configuration_control: ConfigurationControl
    ) -> None:
        """Set configuration control on AirGradient device."""
        await self._set_config("configurationControl", configuration_control)

    async def set_led_bar_mode(self, led_bar_mode: LedBarMode) -> None:
        """Set LED bar mode on AirGradient device."""
        await self._set_config("ledBarMode", led_bar_mode)

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The AirGradientClient object.

        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.

        """
        await self.close()
