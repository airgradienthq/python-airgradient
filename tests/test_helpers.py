"""Asynchronous Python client for Withings."""

from __future__ import annotations

import json
from typing import TYPE_CHECKING, Any

from airgradient import MeasurementGroup, SleepSummary, aggregate_measurements
from airgradient.helpers import aggregate_sleep_summary

from . import load_fixture

if TYPE_CHECKING:
    from syrupy import SnapshotAssertion


def test_aggregate_measurements(snapshot: SnapshotAssertion) -> None:
    """Test aggregation."""
    json_file: list[dict[str, Any]] = json.loads(load_fixture("measurement_list.json"))

    measurements = [MeasurementGroup.from_api(measurement) for measurement in json_file]

    assert aggregate_measurements(measurements) == snapshot


def test_aggregate_sleep_summary(snapshot: SnapshotAssertion) -> None:
    """Test aggregation."""
    json_file: list[dict[str, Any]] = json.loads(load_fixture("sleep_summary.json"))[
        "body"
    ]["series"]

    sleep_summaries = [
        SleepSummary.from_api(sleep_summary) for sleep_summary in json_file
    ]

    assert aggregate_sleep_summary(sleep_summaries) == snapshot


def test_aggregate_single_sleep_summary(snapshot: SnapshotAssertion) -> None:
    """Test aggregation."""
    json_file: dict[str, Any] = json.loads(load_fixture("sleep_summary.json"))["body"][
        "series"
    ][0]

    sleep_summary = SleepSummary.from_api(json_file)

    assert aggregate_sleep_summary([sleep_summary]) == snapshot


def test_aggregate_no_sleep_summary(snapshot: SnapshotAssertion) -> None:
    """Test aggregation."""
    assert aggregate_sleep_summary([]) == snapshot
