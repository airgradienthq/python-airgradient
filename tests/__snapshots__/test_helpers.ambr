# serializer version: 1
# name: test_aggregate_measurements
  dict({
    <MeasurementType.WEIGHT: 1>: 108.4,
    <MeasurementType.HEIGHT: 4>: 1.6,
    <MeasurementType.FAT_FREE_MASS: 5>: 96.0,
    <MeasurementType.FAT_RATIO: 6>: 11.439,
    <MeasurementType.FAT_MASS_WEIGHT: 8>: 12.4,
    <MeasurementType.DIASTOLIC_BLOOD_PRESSURE: 9>: 106.0,
    <MeasurementType.SYSTOLIC_BLOOD_PRESSURE: 10>: 158.0,
    <MeasurementType.HEART_RATE: 11>: 62.0,
    <MeasurementType.MUSCLE_MASS: 76>: 92.2,
    <MeasurementType.HYDRATION: 77>: 70.39,
    <MeasurementType.BONE_MASS: 88>: 3.8000000000000003,
    <MeasurementType.PULSE_WAVE_VELOCITY: 91>: 7.4,
    <MeasurementType.VASCULAR_AGE: 155>: 28.900000000000002,
  })
# ---
# name: test_aggregate_no_sleep_summary
  None
# ---
# name: test_aggregate_single_sleep_summary
  dict({
    'active_movement_duration': None,
    'apnea_hypopnea_index': 9,
    'average_heart_rate': 103,
    'average_movement_score': None,
    'average_respiration_rate': 14,
    'breathing_disturbances_intensity': 9,
    'date': datetime.date(2021, 4, 18),
    'deep_sleep_duration': 5820,
    'end_date': datetime.datetime(2021, 4, 18, 2, 32, 53, tzinfo=datetime.timezone.utc),
    'external_time_asleep': None,
    'hashed_device_id': '201d0b9a0556d6b755166b2cf8d22d3bdf0487ee',
    'light_sleep_duration': 10440,
    'max_heart_rate': 120,
    'max_respiration_rate': 20,
    'min_heart_rate': 70,
    'min_respiration_rate': 10,
    'out_of_bed_count': 0,
    'rem_sleep_duration': 2400,
    'rem_sleep_phase_count': 1,
    'sleep_efficiency': 0.86,
    'sleep_latency': 540,
    'sleep_score': 37,
    'snoring': 1080,
    'snoring_count': 18,
    'start_date': datetime.datetime(2021, 4, 17, 20, 30, 53, tzinfo=datetime.timezone.utc),
    'time_awake_during_sleep': 1380,
    'total_sleep_time': 18660,
    'total_time_awake': 3060,
    'total_time_in_bed': None,
    'wake_up_count': 1,
    'wake_up_latency': 1140,
    'withings_index': None,
  })
# ---
# name: test_aggregate_sleep_summary
  dict({
    'active_movement_duration': None,
    'apnea_hypopnea_index': None,
    'average_heart_rate': 81,
    'average_movement_score': None,
    'average_respiration_rate': 15,
    'breathing_disturbances_intensity': 13,
    'date': datetime.date(2021, 3, 8),
    'deep_sleep_duration': 1492500,
    'end_date': datetime.datetime(2021, 3, 9, 5, 51, 39, tzinfo=datetime.timezone.utc),
    'external_time_asleep': None,
    'hashed_device_id': '201d0b9a0556d6b755166b2cf8d22d3bdf0487ee',
    'light_sleep_duration': 3910680,
    'max_heart_rate': 112,
    'max_respiration_rate': 20,
    'min_heart_rate': 53,
    'min_respiration_rate': 10,
    'out_of_bed_count': None,
    'rem_sleep_duration': 1519620,
    'rem_sleep_phase_count': None,
    'sleep_efficiency': None,
    'sleep_latency': 1017,
    'sleep_score': 92,
    'snoring': 611,
    'snoring_count': 3057,
    'start_date': datetime.datetime(2021, 3, 7, 21, 30, 42, tzinfo=datetime.timezone.utc),
    'time_awake_during_sleep': None,
    'total_sleep_time': None,
    'total_time_awake': 3680,
    'total_time_in_bed': None,
    'wake_up_count': 692,
    'wake_up_latency': 894,
    'withings_index': None,
  })
# ---
